apps/consulting/posts/improvements-to-bsr-matrix-multiplication-in-pytorch.md

```diff
@@ -0,0 +1,302 @@
+axis) crosses the 1.0 threshold, which is marked with a horizontal line labeled "1x". At this point
+the sparse operation executes faster than the dense baseline. This quantity tells us how much
+we need to prune to use sparsity without regressing performance. Looking at the results for
+[`float32`][#fig-2] data type,
+
+<img src="/posts/improvements-to-bsr-matrix-multiplication-in-pytorch/sampled_addmm_float32_vs_dense.png"
+    name="fig-2"
+    alt="A plot displaying sparsity ratio (as a percentage) on the horizontal axis, and speedup over
+    the dense baseline on the vertical.  There is a horizontal line marking the speedup ratio of 1.0
+    above which the sparse subject takes less time to execute than the dense counterpart.  Three lines are
+    plotted indicating block sizes used for the sparse tensor of 16, 32, and 64."
+/>
+
+we see some promising results. Using block sizes of 32 and 64, we cross the 1x threshold at 80%
+sparsity. However, the smallest block size of 16 requires 90% sparsity to pass 1x. The small
+block size is important as it offers more flexibility in the pruning stage as non-zero values can be
+more dispersed, usually resulting in less accuracy loss.
+
+Moving on to [half precision data types][#fig-3], we see much poorer performance compared to dense,
+with the 1x threshold crossed only at 99% sparsity, and only for block size 32 and 64. Requiring
+pruning to this degree is not practical. In most cases, pruning to this level will result in
+unacceptable drops in accuracy.
+
+<img src="/posts/improvements-to-bsr-matrix-multiplication-in-pytorch/sampled_addmm_half_vs_dense.png"
+    name="fig-3"
+    alt="A plot displaying sparsity ratio (as a percentage) on the horizontal axis, and speedup over
+    the dense baseline on the vertical.  There is a horizontal line marking the speedup ratio of 1.0
+    above which the sparse subject takes less time to execute than the dense counterpart.  Three lines are
+    plotted indicating block sizes used for the sparse tensor of 16, 32, and 64."
+/>
+
+These results, albeit promising show there is still room for improvement. Support for
+`sampled_addmm` is a new feature for BSR layout, and we will continue to improve as we work to make
+sparsity a practical technique for transformer acceleration.
+
+**Note**: Although cuSPARSE has added better support for this operation with BSR layout in CUDA 12,
+we do not expose that interface in PyTorch yet.
+
+### `bsr_dense_addmm` (DSD)
+
+The DSD pattern has a direct translation to a dense matrix multiply, so here the experimental setup
+does not require as much explanation. However, support for this type of sparse-dense matrix
+multiplication already exists within PyTorch. In PyTorch 2.0 `BSR @ Dense` multiplication would
+forward to cuSPARSE for float32, and to a custom implementation for half precision types. Here we
+compare to both the dense baseline and these older implementations of the same operation.
```

Comment:
Google flagged this "compare to both". If I change it to "compare against both" it likes it.
idk